### PR TITLE
[NTH-69] feat: therapist appointment table

### PR DESCRIPTION
**Pull request by MohannadSabaa**
- This Branch is for therapist Table Appointments Formed from 2 main Components:

1.  Table With (TIME, BOOKED, AVAILABLE, UNAVAILABLE) fields.
      - Table separated to 2 components (the main table and a row component).
      - Spinner Component.
2. Date Picker.

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -11,7 +11,8 @@ module.exports = {
   parserOptions: { ecmaVersion: 'latest', sourceType: 'module' },
   plugins: ['react-refresh'],
   rules: {
-    'react-refresh/only-export-components': 'warn',
+    'no-nested-ternary': 'warn',
```
Comment (ahmedmurtaja) [inline .eslintrc.cjs]: why editing eslint rules?

File: package.json [hunk 2]
```diff
@@ -17,12 +17,17 @@
     "@material/card": "^14.0.0",
     "@mui/icons-material": "^5.11.16",
     "@mui/material": "^5.13.4",
+    "@mui/x-date-pickers": "^6.8.0",
     "@types/react-router-dom": "^5.3.3",
     "axios": "^1.4.0",
+    "date-fns": "^2.30.0",
+    "dayjs": "^1.11.8",
     "formik": "^2.4.1",
+    "moment": "^2.29.4",
```
Comment (ahmedmurtaja) [inline package.json]: you installed both Moment and dayjs packages, remove unused ones

File: src/components/appointmentsTable/table.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            </TableCell>
+            <TableCell sx={headerCell}>
+              unavailable
+            </TableCell>
+          </TableRow>
+        </TableHead>
+        <TableBody sx={{ margin: '3px', position: 'relative' }}>
+          { loading ? (
+            <TableRow sx={spinner}>
+              <ReactLoading
```
Comment (ahmedmurtaja) [inline src/components/appointmentsTable/table.tsx]: why use React loading, since there is an existing component in MUI check this  [URL](https://mui.com/material-ui/react-progress/)

File: src/components/appointmentsTable/table.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                className="spinner"
+                type="spin"
+                color="#ffe766"
+                width="45px"
+              />
+            </TableRow>
+          ) : (appointments.length
+            ? appointments.map((appo: Appointment) => <RowTable key={appo.id} appo={appo} />) : (
```
Comment (ahmedmurtaja) [inline src/components/appointmentsTable/table.tsx]:
```suggestion
            ? appointments.map((appointment: Appointment) => <RowTable key={appointment.id} appointment={appointment} />) : (

```

File: src/components/appointmentsTable/tableRow.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { useState } from 'react';
+import { TableRow, TableCell, Checkbox } from '@mui/material';
+import { bodyCell } from './style';
+import axiosInstance from '../../utils/apis/axios';
+import { TRow } from './types';
+
+const RowTable = ({ appo }:TRow) => {
```
Comment (ahmedmurtaja) [inline src/components/appointmentsTable/tableRow.tsx]:
```suggestion
const RowTable = ({ appointment }:TRow) => {

```

File: src/components/appointmentsTable/tableRow.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import { TableRow, TableCell, Checkbox } from '@mui/material';
+import { bodyCell } from './style';
+import axiosInstance from '../../utils/apis/axios';
+import { TRow } from './types';
+
+const RowTable = ({ appo }:TRow) => {
+  const {
+    isAvailable, isBooked, datetime, id,
+  } = appo;
+  const [ava, setAva] = useState<boolean>(isAvailable);
```
Comment (ahmedmurtaja) [inline src/components/appointmentsTable/tableRow.tsx]:
```suggestion
  const [available, setAvailability] = useState<boolean>(isAvailable);

```

File: src/components/appointmentsTable/tableRow.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  return (
+    <TableRow sx={{ backgroundColor: ava ? 'white' : '#ffe766' }}>
+      <TableCell sx={bodyCell}>
+        {`${datetime.slice(5, 10)} | ${datetime.slice(12, -8)}`}
+      </TableCell>
+      <TableCell sx={bodyCell}>{isBooked ? 'BOOKED' : 'PENDING'}</TableCell>
+      <TableCell sx={bodyCell}>{ava ? 'AVAILABLE' : 'CANCLED'}</TableCell>
+      <TableCell sx={bodyCell}>
+        <Checkbox
+          disabled={!!isBooked}
```
Comment (ahmedmurtaja) [inline src/components/appointmentsTable/tableRow.tsx]:
``` suggestion
          disabled={isBooked}

```
is booked is already enough

File: package.json [hunk 8]
```diff
@@ -17,12 +17,17 @@
     "@material/card": "^14.0.0",
     "@mui/icons-material": "^5.11.16",
     "@mui/material": "^5.13.4",
+    "@mui/x-date-pickers": "^6.8.0",
     "@types/react-router-dom": "^5.3.3",
     "axios": "^1.4.0",
+    "date-fns": "^2.30.0",
```
Comment (ahmedmurtaja) [inline package.json]: this is unused too

**Review by ahmedmurtaja — APPROVED** [review]

**Review by ahmedmurtaja — CHANGES_REQUESTED** [review]
Great Job Abu Yazid

File: .eslintrc.cjs (hunk 1 above)
Comment (MohannadSabaa) [reply .eslintrc.cjs]: 'react-refresh/only-export-components': 'warn', 
there is something wrong with this rule maybe a syntax one!

Comment (ahmedmurtaja) [reply .eslintrc.cjs]: I am referring to adding this rule 'no-nested-ternary': 'warn',

however, I think its better to keep the configs as default

File: src/components/appointmentsTable/table.tsx (hunk 4 above)
Comment (MohannadSabaa) [reply src/components/appointmentsTable/table.tsx]: it already has a type of Appointment, I think it is enough to be understandable, why to keep just repeating it ?!

File: .eslintrc.cjs (hunk 1 above)
Comment (ahmedmurtaja) [reply .eslintrc.cjs]: and actually you turned off this rule
```'react-refresh/only-export-components': 'warn',```

File: src/components/appointmentsTable/table.tsx (hunk 4 above)
Comment (ahmedmurtaja) [reply src/components/appointmentsTable/table.tsx]: we can request  @LinaYahya or @aaamra to review and make a final choice

Comment (MohannadSabaa) [reply src/components/appointmentsTable/table.tsx]: @raghdarizq she is one of us too

Comment (ahmedmurtaja) [reply src/components/appointmentsTable/table.tsx]: @YasserAmro0 he is one of us too :+1:

Comment (LinaYahya) [reply src/components/appointmentsTable/table.tsx]: I do agree with @ahmedmurtaja ,
Please use clear variable names

File: src/pages/therapistProfile/index.tsx [hunk 9]
```diff
@@ -1,19 +1,22 @@
+import { Container, Button } from '@mui/material';
+import { useParams } from 'react-router-dom';
 import { useState } from 'react';
-import {
-  Button, Container,
-} from '@mui/material';
+import AppointmentTableContainer from '../../components/appointmentsTable';
```
Comment (LinaYahya) [inline src/pages/therapistProfile/index.tsx]: please import from '/compoennts'

File: src/components/appointmentsTable/table.tsx (hunk 3 above)
Comment (LinaYahya) [reply src/components/appointmentsTable/table.tsx]: I do agree, please uninstall react-loading and remove style for `.spinner` and use mui loader

File: src/components/appointmentsTable/index.tsx [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import { Container } from '@mui/material';
+import dayjs, { Dayjs } from 'dayjs';
+import { useState } from 'react';
+import DatePick from './datePicker';
+import AppointmentsTable from './table';
+import { TContainer } from './types';
+
+const AppointmentTableContainer = ({ id }: TContainer) => {
+  const [date, setDate] = useState<Dayjs | string >(dayjs('2023-01-1'));
```
Comment (ahmedmurtaja) [inline src/components/appointmentsTable/index.tsx]: its better that if  the initial value of the state is the current Date

File: src/components/appointmentsTable/index.tsx [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+const AppointmentTableContainer = ({ id }: TContainer) => {
+  const [date, setDate] = useState<Dayjs | string >(dayjs('2023-01-1'));
+  const [loading, setLoading] = useState<boolean>(true);
+
+  return (
+    <Container sx={{ position: 'relative' }}>
+      <DatePick date={date} changeDate={setDate as typeof useState<Dayjs | string>} />
+      <AppointmentsTable
+        id={id}
```
Comment (LinaYahya) [inline src/components/appointmentsTable/index.tsx]: no need to pass id through nested children, I can `useParams` inside `AppointmentsTable`

File: src/components/appointmentsTable/index.tsx (hunk 10 above)
Comment (ahmedmurtaja) [inline src/components/appointmentsTable/index.tsx]:
```suggestion
  const cuurentDate = dayjs().format('YYYY-MM-DD').toString();
  const [date, setDate] = useState<Dayjs | string >(dayjs(cuurentDate));
```

File: src/components/appointmentsTable/index.tsx [hunk 12]
```diff
@@ -0,0 +1,26 @@
+import { Container } from '@mui/material';
+import dayjs, { Dayjs } from 'dayjs';
+import { useState } from 'react';
+import DatePick from './datePicker';
+import AppointmentsTable from './table';
+import { TContainer } from './types';
+
+const AppointmentTableContainer = ({ id }: TContainer) => {
+  const [date, setDate] = useState<Dayjs | string >(dayjs('2023-01-1'));
+  const [loading, setLoading] = useState<boolean>(true);
```
Comment (LinaYahya) [inline src/components/appointmentsTable/index.tsx:10]: why not to move `loading` to `AppointmentsTable` as state it's not sharable at the parent component

File: src/components/appointmentsTable/index.tsx [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+import AppointmentsTable from './table';
+import { TContainer } from './types';
+
+const AppointmentTableContainer = ({ id }: TContainer) => {
+  const [date, setDate] = useState<Dayjs | string >(dayjs('2023-01-1'));
+  const [loading, setLoading] = useState<boolean>(true);
+
+  return (
+    <Container sx={{ position: 'relative' }}>
+      <DatePick date={date} changeDate={setDate as typeof useState<Dayjs | string>} />
```
Comment (LinaYahya) [inline src/components/appointmentsTable/index.tsx]:
```suggestion
      <DatePick date={date} setDate={setDate as typeof useState<Dayjs | string>} />

```

File: src/components/appointmentsTable/table.tsx [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  const [appointments, setAppointments] = useState([]);
+  useEffect(() => {
+    (async () => {
+      try {
+        loadingChange(true);
+        const { data } = await axiosInstance(`/appointments/${id}?date=${date}`);
+        setAppointments(data);
+        loadingChange(false);
+      } catch (e) {
+        loadingChange(false);
```
Comment (LinaYahya) [inline src/components/appointmentsTable/table.tsx]: please render a message incase of unexpected error using `snackbar`

File: src/components/appointmentsTable/renderAppointments.tsx [hunk 15]
```diff
@@ -0,0 +1,19 @@
+import { TableRow, Alert } from '@mui/material';
+import { Appointment } from './types';
+import RowTable from './tableRow';
+import { message } from './style';
+
+const renderAppointments = (appointments: never[]) => (appointments?.length
```
Comment (LinaYahya) [inline src/components/appointmentsTable/renderAppointments.tsx]: It's a more convenint in this wat to create a react component that return 
```jsx
 <RowTable
      key={appointment.id}
      appointment={appointment}
    />
```

and at the parent itself do map through data and handle different cases

File: src/components/appointmentsTable/tableRow.tsx [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+const RowTable = ({ appointment }:TRow) => {
+  const {
+    isAvailable, isBooked, datetime, id,
+  } = appointment;
+  const [availability, setAvailability] = useState<boolean>(isAvailable);
+  return (
+    <TableRow sx={{ backgroundColor: availability ? 'white' : '#ffe766' }}>
+      <TableCell sx={bodyCell}>
+        {`${datetime.slice(5, 10)} | ${datetime.slice(12, -8)}`}
```
Comment (LinaYahya) [inline src/components/appointmentsTable/tableRow.tsx]: we need to convert UTC date to local time, please do your research and find a way using `moemntjs` to do

**Review by LinaYahya — CHANGES_REQUESTED** [review]
Great Job @MohannadSabaa 
A snall nptes to take care of, and please inside `appointmnetsTable` folder try to not having all these files there
Why to combine bot datePicker and table at one place

File: src/index.css [hunk 17]
```diff
@@ -21,6 +21,10 @@ body {
   width: 150px;
   text-align: center;
 }
+.spinner {
```
Comment (LinaYahya) [inline src/index.css]: please remove this

File: src/components/index.tsx [hunk 18]
```diff
@@ -1,4 +1,5 @@
 import AppointmentsModal from './AppointmentsModal';
+import AppointmentTableContainer from './appointmentsTable';
 
 // eslint-disable-next-line import/prefer-default-export
```
Comment (LinaYahya) [inline src/components/index.tsx]: please remove this line

File: .eslintrc.cjs [hunk 19]
```diff
@@ -11,7 +11,8 @@ module.exports = {
   parserOptions: { ecmaVersion: 'latest', sourceType: 'module' },
   plugins: ['react-refresh'],
   rules: {
-    'react-refresh/only-export-components': 'warn',
+    'no-nested-ternary': 'off',
+    'react-refresh/only-export-components': 2,
```
Comment (LinaYahya) [inline .eslintrc.cjs]: please revert these changes

File: src/components/appointmentsTable/table/index.tsx [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+          margin: '3px', position: 'relative',
+        }}
+        >
+          { loading
+            && (
+            <TableRow sx={spinner}>
+              <CircularProgress />
+            </TableRow>
+            ) }
+          {appointments
```
Comment (LinaYahya) [inline src/components/appointmentsTable/table/index.tsx]: check loading before

File: src/components/appointmentsTable/table/index.tsx [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+            && appointments?.map(
+              (appointment: Appointment) => (
+                <RowTable
+                  key={appointment.id}
+                  appointment={appointment}
+                />
+              ),
+            )}
+          {appointments?.length === 0
```
Comment (LinaYahya) [inline src/components/appointmentsTable/table/index.tsx]: check loading before

File: src/components/appointmentsTable/table/tableRow.tsx [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+      </TableCell>
+      <TableCell sx={bodyCell}>{isBooked ? 'BOOKED' : 'PENDING'}</TableCell>
+      <TableCell sx={bodyCell}>{availability ? 'AVAILABLE' : 'CANCELED'}</TableCell>
+      <TableCell sx={bodyCell}>
+        <Checkbox
+          disabled={isBooked}
+          checked={!availability}
+          onChange={async () => {
+            setAvailability(!availability);
+            await axiosInstance.put(`/appointments/${id}`);
```
Comment (LinaYahya) [inline src/components/appointmentsTable/table/tableRow.tsx]: please try to check if a request is success then setAvailability

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

**Review by YasserAmro0 — APPROVED** [review]

**Review by ahmedmurtaja — APPROVED** [review]

**Review by raghdarizq — APPROVED** [review]
